Flink: if provided, add uidPrefix to operator name so that Flink web …

stevenzwu opened the pull request:
…UI can show different names when there are multiple Iceberg sinks for a Flink job

openinx commented:
Thanks @stevenzwu for the PR,   Let me take a look...

openinx commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -209,21 +209,23 @@ public Builder equalityFieldColumns(List<String> columns) {
      * pipeline.auto-generate-uids=false} to disable auto-generation and force explicit setting of all operator uids.
-     * <p>
+     * <br><br>
      * Be careful with setting this for an existing job, because now we are changing the opeartor uid from an
```
`opeartor` ?  typo ?

openinx commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -246,52 +248,71 @@ public Builder uidPrefix(String newPrefix) {
-      IcebergStreamWriter<RowData> streamWriter = createStreamWriter(table, flinkRowType, equalityFieldIds);
-      IcebergFilesCommitter filesCommitter = new IcebergFilesCommitter(tableLoader, overwrite);
+      // Add parallel writers that append rows to files
+      SingleOutputStreamOperator<WriteResult> writerStream = appendWriter(rowDataInput, flinkRowType);
```
The [unchecked](https://github.com/apache/iceberg/pull/2886/files#diff-a26cb8b3f0f5b0e6a9f110bb0b3fdd0ed625bfc9faf27e248f6446818cad4d69L234) warning can be removed now ?

openinx commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -246,52 +248,71 @@ public Builder uidPrefix(String newPrefix) {
-          .addSink(new DiscardingSink())
-          .name(String.format("IcebergSink %s", table.name()))
-          .setParallelism(1);
+    private SingleOutputStreamOperator<WriteResult> appendWriter(DataStream<RowData> input, RowType flinkRowType) {
```
Shouldn't it break any unit test that address [distributeDataStream](https://github.com/apache/iceberg/pull/2886/files#diff-a26cb8b3f0f5b0e6a9f110bb0b3fdd0ed625bfc9faf27e248f6446818cad4d69R255) logic  ?  I see we did not use the `input` data stream  when chain the iceberg writers into the stream..

stevenzwu replied:
we are still reference to the stream after `distributeDataStream`. still using the old style of mutable `rowDataInput`. I definitely see that it is easier to get confused. let me update the code to make it more consistent and clear

stevenzwu replied:
after re-read the code, I saw your point now. will fix the `input` references

openinx commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -246,52 +248,71 @@ public Builder uidPrefix(String newPrefix) {
+    }
+
+    private DataStreamSink<RowData> appendDummySink(SingleOutputStreamOperator<Void> committerStream) {
+      DataStreamSink<RowData> resultStream = committerStream
+          .addSink(new DiscardingSink())
+          .name(getOperatorName(String.format("IcebergSink %s", this.table.name())))
+          .setParallelism(1);
       if (uidPrefix != null) {
-        writerStream = writerStream.uid(uidPrefix + "-writer");
+        resultStream = resultStream.uid(uidPrefix + "-dummysink");
       }
+      return resultStream;
+    }
```
Seems like the correct code should be the following, so that we could get ride of the IDE warning:

```java
    private DataStreamSink<Void> appendDummySink(SingleOutputStreamOperator<Void> committerStream) {
      DataStreamSink<Void> resultStream = committerStream
          .addSink(new DiscardingSink<>())
          .name(getOperatorName(String.format("IcebergSink %s", this.table.name())))
          .setParallelism(1);
      if (uidPrefix != null) {
        resultStream = resultStream.uid(uidPrefix + "-dummysink");
      }
      return resultStream;
    }
```

stevenzwu replied:
You are right here. I see the problem is the build method
```
public DataStreamSink<RowData> build()
```

openinx replied:
I think the `build()` will need to be changed to return a `DataStreamSink<Void>` because the stream won't actually emit any real `RowData`.  But as it's a public API,  will this break the upstream flink datastream job?

openinx commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -246,52 +248,71 @@ public Builder uidPrefix(String newPrefix) {
+      return appendDummySink(committerStream);
+    }
+
+    private String getOperatorName(String suffix) {
```
In iceberg, we usually remove the `get` prefix from the getter method...

openinx reviewed (changes requested):
Thanks @stevenzwu  for the PR to address the multi-sinks in the same flink job,  I left several comments.  I think we need at least one unit test addressing the multi-sink in same flink job so that we won't regress in future.

stevenzwu commented:
@openinx I addressed your comments in the last commit. it is a good call to add a unit test. Through the unit test, I also found that uid wasn't set the map operator. Latest commit addressed it. Because I don't want to change the `Builder` class to a generic type. I added a `mappedRowDataInput` variable to distinguish this case so that we can add the uid and name in the `build()` method.

openinx commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -135,6 +135,13 @@ private Builder forRowData(DataStream<RowData> newRowDataInput) {
     }
 
+    private <T> Builder forMapperOutputType(DataStream<T> input,
+                                        MapFunction<T, RowData> mapper,
```
Nit: missing a code format.

stevenzwu replied:
thx. will fix (along with any other changes if needed)

openinx replied:
Yeah,  this is needed.

stevenzwu replied:
fixed

openinx commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -118,6 +117,7 @@ public static Builder forRowData(DataStream<RowData> input) {
 
   public static class Builder {
     private DataStream<RowData> rowDataInput = null;
+    private SingleOutputStreamOperator<RowData> mappedRowDataInput = null;
```
I'd prefer not introducing another `DataStream<RowData>` named `mappedRowDataInput` into this Builder,  because that could easily confuse people:  which `DataStream` should I use to chain the following operators ?  Let's just keep the only one `rowDataInput` even if we've chained a `MapFunction`,  we could still add the `uid` and `name` lazily to the `Map` operator ?

stevenzwu replied:
Note that `mappedRowDataInput` is intentionally defined as `SingleOutputStreamOperator` (not `DataStream `), because `DataStream` doesn't expose `uid` and `name` method and hence we can't lazily add them to the `Map` operator with `DataStream<RowData> rowDataInput`.

I agree that this is a little confusing and less than ideal. Another option is to change Builder to a generic class like `Builder<T>`. but I don't like that either and hence used the current implementation. I am very open to other alternatives.

openinx replied:
I tried to initialize the input stream lazily as the following because I think it could make the source `rowDataInput` stream very clear, it's the only one stream that we will need to chain the following operator.

```diff
diff --git a/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java b/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java
index bda6667cd..a6be26b31 100644
--- a/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java
+++ b/flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java
@@ -24,6 +24,7 @@ import java.io.UncheckedIOException;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.function.Function;
 import org.apache.flink.api.common.functions.MapFunction;
 import org.apache.flink.api.common.typeinfo.TypeInformation;
 import org.apache.flink.api.common.typeinfo.Types;
@@ -117,7 +118,7 @@ public class FlinkSink {
 
   public static class Builder {
     private DataStream<RowData> rowDataInput = null;
-    private SingleOutputStreamOperator<RowData> mappedRowDataInput = null;
+    private Function<String, DataStream<RowData>> inputCreator = null;
     private TableLoader tableLoader;
     private Table table;
     private TableSchema tableSchema;
@@ -136,9 +137,17 @@ public class FlinkSink {
     }
 
     private <T> Builder forMapperOutputType(DataStream<T> input,
-                                        MapFunction<T, RowData> mapper,
-                                        TypeInformation<RowData> outputType) {
-      this.mappedRowDataInput = input.map(mapper, outputType);
+                                            MapFunction<T, RowData> mapper,
+                                            TypeInformation<RowData> outputType) {
+      this.inputCreator = newUidPrefix -> {
+        if (newUidPrefix != null) {
+          return input.map(mapper, outputType)
+              .name(operatorName(newUidPrefix))
+              .uid(newUidPrefix + "-mapper");
+        } else {
+          return input.map(mapper, outputType);
+        }
+      };
       return this;
     }
 
@@ -241,8 +250,11 @@ public class FlinkSink {
     }
 
     public DataStreamSink<RowData> build() {
-      Preconditions.checkArgument(rowDataInput != null || mappedRowDataInput != null,
-          "Please use forRowData() or forMapperOutputType() to initialize the input DataStream.");
+      Preconditions.checkArgument(rowDataInput == null || inputCreator == null,
+          "Use forRowData() or forRow() to initialize the input DataStream only once.");
+      Preconditions.checkArgument(rowDataInput != null || inputCreator != null,
+          "Please use forRowData() or forRow() to initialize the input DataStream.");
+      rowDataInput = rowDataInput == null ? inputCreator.apply(uidPrefix) : rowDataInput;
       Preconditions.checkNotNull(tableLoader, "Table loader shouldn't be null");
 
       if (table == null) {
@@ -254,17 +266,6 @@ public class FlinkSink {
         }
       }
 
-      // set name and uid for mappedRowDataInput if needed
-      if (mappedRowDataInput != null) {
-        if (uidPrefix != null) {
-          rowDataInput = mappedRowDataInput
-              .name(operatorName(uidPrefix))
-              .uid(uidPrefix + "-mapper");
-        } else {
-          rowDataInput = mappedRowDataInput;
-        }
-      }
-
       // Convert the requested flink table schema to flink row type.
       RowType flinkRowType = toFlinkRowType(table.schema(), tableSchema);
```

stevenzwu replied:
great idea. switched to this lazy ini mode

openinx commented on flink/src/main/java/org/apache/iceberg/flink/sink/FlinkSink.java:
```diff
@@ -246,52 +254,83 @@ public Builder uidPrefix(String newPrefix) {
       // Distribute the records from input data stream based on the write.distribution-mode.
-      rowDataInput = distributeDataStream(rowDataInput, table.properties(), table.spec(), table.schema(), flinkRowType);
+      DataStream<RowData> distributeStream = distributeDataStream(
+          rowDataInput, table.properties(), table.spec(), table.schema(), flinkRowType);
```
Should we also add the uid & name if we've already chained the `PartitionKeySelector` ?

stevenzwu replied:
`keyBy` returns a `KeyedStream` which extends from `DataStream` that doesn't expose `name` and `uid`

openinx commented on flink/src/test/java/org/apache/iceberg/flink/sink/TestFlinkIcebergSink.java:
```diff
@@ -258,4 +262,58 @@ public void testShuffleByPartitionWithSchema() throws Exception {
+  public void testTwoSinksInDisjointedDAG() throws Exception {
+    Map<String, String> props = ImmutableMap.of(TableProperties.DEFAULT_FILE_FORMAT, format.name());
+
+    final String leftTablePath = TEMPORARY_FOLDER.newFolder().getAbsolutePath().concat("/left");
```
Unit tests looks good to me.  One minor thing:  in apache iceberg codebase,   I have almost never seen a situation where local variables are declared as final, @rdblue , Is this an implicit code style rule?

stevenzwu replied:
done. I removed all the final modifiers for local vars in this PR

openinx reviewed (commented):
@stevenzwu , patch looks good to me overall.  I think the left comments will need to be fixed.  Thanks for the patch !

stevenzwu commented:
@openinx please take another look. Thanks!

openinx reviewed (approved):
LGTM now, thanks @stevenzwu !
